app/models/order.rb

```diff
@@ -0,0 +1,151 @@
+# frozen_string_literal: true
+
+class Order < ApplicationRecord
+  include AASM
+
+  has_many   :addresses, as: :addressable, dependent: :destroy
+  has_many   :order_items, dependent: :destroy, after_add: :recalculate_total, after_remove: :recalculate_total
+  has_many   :books, through: :order_items
+  has_one    :coupon, dependent: :destroy
+  belongs_to :card, dependent: :destroy
+  belongs_to :user
+  belongs_to :delivery
+
+  before_create :set_track_number
+
+  validates_length_of :track_number, maximum: 25
+
+  MY_ORDERS_STATES = %i(in_confirmation in_processing in_delivery completed).freeze
+
+  scope :in_carting, -> { where(aasm_state: %i(cart filled)) }
+  scope :after_cart, -> { where(aasm_state: MY_ORDERS_STATES) }
+  scope :newest, -> { order('created_at DESC') }
+  scope :active, -> { where.not(aasm_state: :canceled) }
+
+  aasm column: 'aasm_state' do
+    state :cart, initial: true
+    state :filled
+    state :in_confirmation, after_enter: :send_confirmation
+    state :in_processing, after_enter: :treat_proccessing
+    state :in_delivery
+    state :completed, after_enter: :send_success
+    state :canceled
+
+    event :filled do
+      transitions from: :cart, to: :filled
+    end
+
+    event :in_confirmation do
+      transitions from: :filled, to: :in_confirmation
+    end
+
+    event :treat do
+      transitions from: :in_confirmation, to: :in_processing
+    end
+
+    event :to_deliver do
+      transitions from: :in_processing, to: :in_delivery
+    end
+
+    event :complete do
+      transitions from: :in_delivery, to: :completed
+    end
+
+    event :cancel do
+      transitions from: %i(cart filled in_confirmation in_delivery in_processing completed), to: :canceled
+    end
+  end
+
+  rails_admin do
+    list do
+      field :aasm_state, :state
+      include_all_fields
+      exclude_fields :confirmation_token, :completed_date
+    end
+    edit do
+      field :aasm_state, :state
+      include_all_fields
+      exclude_fields :confirmation_token, :completed_date
+    end
+  end
+
+  def valid_addresses?
+    shipping_address && billing_address
+  end
+
+  def self.default_sort
+    ORDERING[DEFAULT_SORT_KEY]
+  end
+
+  def billing_address
+    addresses.find_by(kind: 'billing')
+  end
+
+  def shipping_address
+    addresses.find_by(kind: 'shipping')
+  end
+
+  def book_in_order?(book_id)
+    !order_items.find_by(book: book_id).nil?
+  end
+
+  def subtotal_more_than_discount?(coupon)
+    subtotal_price > coupon.discount
+  end
+
+  def send_confirmation
+    update_attributes(confirmation_token: Devise.friendly_token)
+    OrderMailer.confirmation_send(user, self).deliver_now
```

Comment:
When you use update_attributes save is calling so you can use before/after save callback with observer and check if confirmation_token was changed. It is just possible variant :)